src/reduce_dims.cpp

```diff
@@ -0,0 +1,117 @@
+#include <migraphx/reduce_dims.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool reduce_dim(std::vector<shape>& shapes, std::size_t n)
+{
+    std::vector<std::size_t> new_lens;
+    for(const auto& s : shapes)
+    {
+        if(n >= s.lens().size())
+            return false;
+        if((n + 1) >= s.lens().size())
+            return false;
+        auto astride = s.strides()[n];
+        auto alen    = s.lens()[n];
+        auto bstride = s.strides()[n + 1];
+        auto blen    = s.lens()[n + 1];
+
+        if(astride == bstride * blen)
+        {
+            new_lens.push_back(alen * blen);
+        }
+    }
+    if(new_lens.size() != shapes.size())
+        return false;
+    std::size_t i = 0;
+    for(auto& s : shapes)
+    {
+        auto lens    = s.lens();
+        auto strides = s.strides();
+        lens.erase(lens.begin() + n);
```

Comment:
No, this is only collapsing two dimensions. It also possible that only the middle dimension can be collapsed. It is probably a good idea to add a test case for that.

Also, since we are collapsing `lens[n]` and `lens[n+1]` we can erase either one. I just erase the first one. This also makes consistent when collapsing the strides because for strides, `stride[n]` should be the value of `strides[n+1]` before collapsing.